core/src/evaluation/functions/future/previous_value.rs

```
@@ -0,0 +1,184 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::BTreeMap;
+use std::sync::Arc;
+use std::sync::Weak;
+
+use crate::evaluation::context::SideEffects;
+use crate::evaluation::functions::aggregation::ValueAccumulator;
+use crate::evaluation::functions::ScalarFunction;
+use crate::evaluation::variable_value::VariableValue;
+use crate::evaluation::ExpressionEvaluationContext;
+use crate::evaluation::ExpressionEvaluator;
+use crate::evaluation::{FunctionError, FunctionEvaluationError};
+use crate::interface::ResultIndex;
+use crate::interface::ResultKey;
+use crate::interface::ResultOwner;
+use crate::models::ElementValue;
+use async_trait::async_trait;
+use drasi_query_ast::ast;
+
+pub struct PreviousValue {
+    result_index: Arc<dyn ResultIndex>,
+    expression_evaluator: Weak<ExpressionEvaluator>,
+}
+
+impl PreviousValue {
+    pub fn new(
+        result_index: Arc<dyn ResultIndex>,
+        expression_evaluator: Weak<ExpressionEvaluator>,
+    ) -> Self {
+        Self {
+            result_index,
+            expression_evaluator,
+        }
+    }
+}
+
+#[async_trait]
+impl ScalarFunction for PreviousValue {
+    async fn call(
+        &self,
+        context: &ExpressionEvaluationContext,
+        expression: &ast::FunctionExpression,
+        args: Vec<VariableValue>,
+    ) -> Result<VariableValue, FunctionError> {
+        if args.is_empty() {
+            return Err(FunctionError {
+                function_name: expression.name.to_string(),
+                error: FunctionEvaluationError::InvalidArgumentCount,
+            });
+        }
+
+        let result_owner = ResultOwner::Function(expression.position_in_query);
+        let value = &args[0];
+        let default_value: ElementValue = {
+            if args.len() > 1 {
```

nit: maybe we should throw an error if the args.len() is an invalid number